ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -528,12 +545,42 @@ RouterException getChunkException() {
      *                                    created as part of this poll operation.
      */
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
-      //First, check if any of the existing requests have timed out.
-      cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!maybeProcessDecryptCallbackAndCompleteOperation()) {
+        //First, check if any of the existing requests have timed out.
+        cleanupExpiredInFlightRequests();
+        checkAndMaybeComplete();
+        if (!isComplete() && !onDecryptMode()) {
+          fetchRequests(requestRegistrationCallback);
+        }
+      }
+    }
+
+    /**
+     * Maybe process decrypt callback and complete the operation if applicable. This is a no-op for non-encrypted blobs
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    protected boolean maybeProcessDecryptCallbackAndCompleteOperation() {
+      if (onDecryptMode() && decryptCallbackResultInfo.get().resultAvailable) {
+        logger.trace("Processing decrypt callback stored result for data chunk {}", chunkBlobId);
+        DecryptCallBackResultInfo resultInfo = decryptCallbackResultInfo.get();
+        if (resultInfo.exception == null && (getChunkException() == null || !retainChunkExceptionOnSuccess)) {
+          chunkIndexToBuffer.put(chunkIndex, filterChunkToRange(resultInfo.result.getDecryptedBlobContent()));
+          numChunksRetrieved++;
+          logger.trace("Decrypt result successfully updated for data chunk {}", chunkBlobId);
+        } else {
+          logger.trace("Setting operation exception as decryption callback invoked with exception {} for data chunk {}",
+              resultInfo.exception, chunkBlobId);
+          setOperationException(
+              new RouterException("Exception thrown on decrypting the content for data chunk " + chunkBlobId,
+                  resultInfo.exception, RouterErrorCode.UnexpectedInternalError));
+        }
+        logger.trace("Marking blob content available to process for data chunk {}", chunkBlobId);
+        blobContentAvailableToProcess = true;
+        checkAndMaybeComplete();
```

Comment:
this seems to get called from the poll anyway?

Reply:
yes. But usual flow to complete an operation is handleReponse -> handleBody -> checkAndMaybeComplete -> onChunkOperationComplete(). 

In poll(), we usually don't call onChunkOperationComplete(). So, I'd like to maintain that and not introduce changes. Hence, maybeProcessDecryptCallbackAndCompleteOperation() is solely responsible for completing the operation as in the normal flow above i.e. checkAndMaybeComplete -> onChunkOperationComplete().